### PR TITLE
Monitor proper cell

### LINKED ISSUE #14: Monitoring appears under wrong cell
When I use shortcut Shift + Enter to run the current cell (C0) and select below one (C1), monitoring appears under currently selected cell (C1), but I expect to see it under C0.

### PR DESCRIPTION
**Pull request by Ben-Epstein**
Fixes #14
Putting the cell monitor in the correct code cell and support for multiple spark sessions. The Spark UI is now at `URL/sparkmonitor/PORT` where `PORT` is the port of the spark UI (ie `/sparkmonitor/4040`, `/sparkmonitor/4041` etc

### CODE REVIEW

File: js/CellMonitor.js [hunk 1]
```diff
@@ -24,7 +24,8 @@ export default class CellMonitor {
      */
     constructor(monitor, cell) {
         this.monitor = monitor; // Parent SparkMonitor instance
-        this.cell = cell; // Jupyter Cell instance
+        // this.cell = cell; // Jupyter Cell instance
+        this.cell = cell
```
Comment (itsjafer) [inline js/CellMonitor.js]: Could we clean up this line so we don’t have a lingering comment on line 27?

File: js/SparkMonitor.js [hunk 2]
```diff
@@ -176,26 +189,31 @@ export default class SparkMonitor {
      * Called when a Spark job is started.
      * @param {Object} data - The data from the spark listener event.
      */
-    onSparkJobStart(data) {
-        const cell = this.listener.getActiveCell();
+    onSparkJobStart(data) {
```
Comment (itsjafer) [inline js/SparkMonitor.js]: nit pick: could we get rid of the extra space here

File: package.json [hunk 3]
```diff
@@ -78,4 +78,4 @@
     "js/*",
     "dist/*"
   ]
-}
+}
```
Comment (itsjafer) [inline package.json]: Could we add the new line back to the end of this file?

**Review by itsjafer — CHANGES_REQUESTED** [review]
Thank you so much for this addition! Skimming through the code, I really only see changed needed in formatting and style. Could you install the pre-commit hooks (https://pre-commit.com/) as well as run eslint on the javascript files to iron out the little things? Thank you!

**Comment by Ben-Epstein** [comment]
@itsjafer I addressed the comments, thanks for the review. Sorry, I haven't used https://pre-commit.com/ or eslint before. What do I need to run?

**Update** I ran the pre-commit and reformatted the code. Still not sure what eslint is.

**Comment by itsjafer** [comment]
> @itsjafer I addressed the comments, thanks for the review. Sorry, I haven't used https://pre-commit.com/ or eslint before. What do I need to run?

I saw you already ran the pre-commit which is awesome. To run eslint, you can run `yarn eslint path/to/javascript/file/you/changed.js --fix` which will use the eslint config in the root of the project folder to lint the files you changed. You could alternatively run eslint from within the node_modules folder: `node_modules/eslint/bin/eslint.js path/to/javascript/file/you/changed.js --fix`

In the future, I'm gonna update the readme to include these instructions!

**Comment by Ben-Epstein** [comment]
@itsjafer finished eslint. Please try to rebuild and test if you haven't already before merge :) 

**Update:** I can no longer build the extension from my deployment [jupyterlab_sparkmonitor-s@1.1.4](https://www.npmjs.com/package/jupyterlab_sparkmonitor-s). Local build is working fine. 

Maybe I'm packaging and deploying it wrong?

Recognize anything in this error?
[err.log](https://github.com/itsjafer/jupyterlab-sparkmonitor/files/4957044/err.log)

**Comment by itsjafer** [comment]
> @itsjafer finished eslint. Please try to rebuild and test if you haven't already before merge :)
> 
> **Update:** I can no longer build the extension from my deployment [jupyterlab_sparkmonitor-s@1.1.4](https://www.npmjs.com/package/jupyterlab_sparkmonitor-s). Local build is working fine.
> 
> Maybe I'm packaging and deploying it wrong?
> 
> Recognize anything in this error?
> [err.log](https://github.com/itsjafer/jupyterlab-sparkmonitor/files/4957044/err.log)

It seems to me the problem is that the eslint rule conflicts with whatever web pack rules jupyterlab uses. It's been a while, but I believe the way I get around that is by setting the entry point in package.json line 74 to "lib/index" which is where babel compiles the code.

I also ran your code locally on my machine and it compiled fine but I wasn't actually able to get the CellMonitor to appear in the correct cell. When I ran your example notebook, the monitor seemed to appear in the bottom cell. When I created a new notebook with the cells below, I wasn't able to get a cell monitor to appear at all! The console reported that `SparkMonitor: Stage started with no running cell.`:

image

Could you try running `git clean -ffdx` and then recompiling the extension and testing the notebook in the screenshot?

**Comment by Ben-Epstein** [comment]
@itsjafer I ran the clean
Screen Shot 2020-07-21 at 10 34 27 PM

Then ran 
```
make venv
make build
make develop
```

I took a screen recording but only gifs are supported so the quality is shot. Here's a screenshot too 
Screen Shot 2020-07-21 at 10 43 44 PM

sparkmonitor


I also tried your change in package.json. It's a new error now
```
yarn run v1.21.1
$ ensure-max-old-space webpack --config webpack.prod.minimize.config.js
ModuleNotFoundError: Module not found: Error: Can't resolve './styles.css' in '/opt/conda/share/jupyter/lab/staging/node_modules/jupyterlab_sparkmonitor-s/lib'
```

**Comment by itsjafer** [comment]
@Ben-Epstein I wonder why there's a discrepancy; could it be that we're running different Jupyterlab versions? Here's what happens when I pull your branch unchanged, run the development commands, and execute your example notebook (removing some of the extra cells to fit the whole thing in the recording). It's pretty odd how the monitor appears at the random cell at the bottom; I tried messing around with executing different cells and using different shortcuts to execute and had the same result. For reference, I'm on jupyterlab 2.2.0!

*Edit:* I also noticed you didn't source the venv (`source venv/bin/activate`), maybe you're using a global version of jupyterlab instead of the version installed in the virtual environment?


Screen Recording 2020-07-22 at 1 07 15 AM

I'll also take a look through the code and try to see if its a problem on my end!

**Comment by Ben-Epstein** [comment]
@itsjafer This is really strange. I source activated my venv. I created and deleted cells. I moved cells around during execution.. No issues. 


sparkmonitor2



Screen Shot 2020-07-22 at 9 17 11 AM


Screen Shot 2020-07-22 at 9 26 27 AM


Are you running linux or macos?

**Comment by Ben-Epstein** [comment]
Fixed and PR ready based on our call. The override isn't set yet. You still need to set the `{"recordTiming": True}` in the advanced notebook settings (for now).

File: setup.py [hunk 4]
```diff
@@ -30,6 +31,6 @@
       install_requires=[
           'bs4',
           'tornado',
-          'pyspark'
+          'pyspark==2.4.3'
```
Comment (itsjafer) [inline setup.py]: Is this a necessary change?

File: package.json [hunk 5]
```diff
@@ -15,7 +15,7 @@
   ],
   "author": "Krishnan R",
   "maintainer": "itsjafer",
-  "license": "LGPL-2.1",
+  "license": "Apache 2.0",
```
Comment (itsjafer) [inline package.json]: I get a warning from npm because this isn't a valid license: I think it should be 'Apache-2.0'.

**Review by itsjafer — APPROVED** [review]
Looks good to me; I had two small nit picks but the PR is approved right now so you can merge when those changes are made. 

This change relies on JupyterLab 2.0+ which means this is a breaking change (since the extension currently supports Jupyterlab 1 and 2); after this is merged into master, I'll update the readme with instructions about setting user-level settings and probably publish this with a major version update to 2.0.

**Comment by Ben-Epstein** [comment]
@itsjafer fixed, but i don't have permissions to merge. Whenever you're ready :)

File: package.json [hunk 6]
```diff
@@ -1,6 +1,6 @@
 {
-  "name": "jupyterlab_sparkmonitor",
-  "version": "1.1.0",
+  "name": "jupyterlab_sparkmonitor-s",
+  "version": "1.2.0",
```
Comment (itsjafer) [inline package.json]: Whoops @Ben-Epstein, I think this change accidentally slipped in. Could we change the name and version back to their original values? The name probably was an accidental change and I'll update the version in master when I publish :)

**Review by itsjafer — APPROVED** [review]
Sorry for the tiny nit pick!

File: package.json (hunk 6 above)
Comment (Ben-Epstein) [reply package.json]: done

**Pull request merged**